specs/UnhandledKeyPressed.md

```diff
@@ -0,0 +1,248 @@
+        (e.Modifiers & CoreWebView2KeyPressedFlagKind.CoreWebView2KeyEventFlagControlDown) != 0)
+    {
+        if (e.VirtualKey == 'Z')
+        {
+            // Raised when not focused in an edit control, or
+            // there is nothing to undo in an edit control.
+            MessageBox.Show("Key combination Ctrl + Z unhandled by browser is triggered.");
+        }
+        else if (e.VirtualKey >= 'A' && e.VirtualKey <= 'Z')
+        {
+            Debug.WriteLine($"Ctrl + {(char)e.VirtualKey} not handled by browser is triggered.");
+        }
+    }
+}
+```
+
+# API Details
+
+## Win32 C++
+```idl
+[uuid(053b9a5d-7033-4515-9898-912977d2fde8), object, pointer_default(unique)]
+interface ICoreWebView2Controller : IUnknown {
+  /// Adds an event handler for the `UnhandledKeyPressed` event.
+  /// `UnhandledKeyPressed` will be raised ONLY IF ALL of the following conditions are met:
+  /// 1. The key event has not been marked as handled in the 'AcceleratorKeyPressed' event.
+  /// 2. The key event is not consumed in the DOM (e.g., through calling event.preventDefault()).
+  /// 3. The event is not recognized or taken over by standard browser shortcuts (like ctrl + F, ctrl + P, etc).
+
+  HRESULT add_UnhandledKeyPressed(
+    [in] ICoreWebView2UnhandledKeyPressedEventHandler* eventHandler,
+    [out] EventRegistrationToken* token);
+
+  /// Removes an event handler previously added with
+  /// `add_UnhandledKeyPressed`.
+
+  HRESULT remove_UnhandledKeyPressed(
+    [in] EventRegistrationToken token);
+}
+
+[uuid(dc83113b-ce7a-47bb-9661-16b03ff8aac1), object, pointer_default(unique)]
+interface ICoreWebView2UnhandledKeyPressedEventHandler : IUnknown {
+
+  /// Provides the event args for the corresponding event.
+
+  HRESULT Invoke(
+      [in] ICoreWebView2Controller* sender,
+      [in] ICoreWebView2UnhandledKeyPressedEventArgs* args);
+}
+
+[uuid(dd30e20c-d1b3-4e51-9bb6-1be7aaa3ce90), object, pointer_default(unique)]
+interface ICoreWebView2UnhandledKeyPressedEventArgs : IUnknown {
+
+  /// The key event type that caused the event to run.
+
+  [propget] HRESULT KeyEventKind([out, retval] COREWEBVIEW2_KEY_EVENT_KIND* keyEventKind);
+
+  /// The Win32 virtual key code of the key that was pressed or released.  It
+  /// is one of the Win32 virtual key constants such as `VK_RETURN` or an
+  /// (uppercase) ASCII value such as `A`.
+  /// use ICoreWebView2UnhandledKeyPressedEventArgs::get_Modifiers()
+  /// instead of Win32 API ::GetKeyState() to verify whether Ctrl or Alt
+  /// is pressed.
+
+  [propget] HRESULT VirtualKey([out, retval] UINT* virtualKey);
+
+  /// The `LPARAM` value that accompanied the window message.  For more
+  /// information, navigate to [WM_KEYDOWN](/windows/win32/inputdev/wm-keydown)
+  /// and [WM_KEYUP](/windows/win32/inputdev/wm-keyup).
+  /// For visual hosting, only scan code and extended key (16-24bit) are provided.
+  /// For other bits, check out the origin message from ::GetMessage().
```

Comment:
This seems to imply that the app can call GetMessage to retrieve the information. Even if the app saved the MSG from the call to GetMessage, how would they match that to an UnhandledKeyPressed event since the event is async from the MSG being received?

Reply:
Devs should expect which keys should fire UnhandledKeyPressed by themselves to know which message the UnhandledKeyPressed is related to. It is not very healthy. I did not expect it to be used often, the cases they need to do it is when they need the repeat count and previous key state according to the [doc](https://learn.microsoft.com/en-us/windows/win32/inputdev/wm-keydown). Maybe we can make some changes to on how key events are sent to browser for visual hosting to avoid this.

Reply:
Yes, what I mean was the dev should keep everything including how many characters users are putting in a textbox -- that includes lot of work on the js side and it is not healthy. 
After weighing the use case and the cost of modification, I decided not to change the event and added a comment. Fixing this by making some modification to the event is fine to me.